source/opt/ilist_node.h

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIBSPIRV_OPT_ILIST_NODE_H_
+#define LIBSPIRV_OPT_ILIST_NODE_H_
+
+#include <cassert>
+
+namespace spvtools {
+namespace utils {
+
+template <class NodeType>
+class IntrusiveList;
+
+// IntrusiveNodeBase is the class when creating classes than can be inserted
+// into an IntrusiveList.  See the comments in ilist.h on how to use the class.
+
+template <class NodeType>
+class IntrusiveNodeBase {
+ public:
+  // Creates a new node that is not in a list.
+  inline IntrusiveNodeBase();
+
+  // Returns the node that comes after the given node in the list, if one
+  // exists.  If the given node is not in a list or is at the end of the list,
+  // the return value is nullptr.
+  inline NodeType* NextNode() const;
+
+  // Returns the node that comes before the given node in the list, if one
+  // exists.  If the given node is not in a list or is at the start of the
+  // list, the return value is nullptr.
+  inline NodeType* PreviousNode() const;
+
+  // Will insert the given node immediately before pos in the list.
+  // This the given node is already in a list, it will first be removed
+  // from that list.
+  //
+  // It is assumed that the given node is of type NodeType.  It is an error is
+  // pos is not already in a list.
+  inline void InsertBefore(NodeType* pos);
+
+  // Will insert the given node immediately after pos in the list.
+  // This the given node is already in a list, it will first be removed
+  // from that list.
+  //
+  // It is assumed that the given node is of type NodeType.  It is an error is
+  // pos is not already in a list.
+  inline void InsertAfter(NodeType* pos);
+
+  // Removes the given node from the list.  Note that this does not free any
+  // storage related to the node.
+  inline void RemoveFromList();
+
+ private:
+  NodeType* next_node_;
+  NodeType* previous_node_;
+  bool is_sentinel_;
```

Comment:
// True only for the sentinel node stored in the list itself.

Reply:
done